soc/arm/atmel_sam0/common/soc_gpio.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2016 Piotr Mienkowski
+ * Copyright (c) 2018 Google LLC.
+ * Copyright (c) 2021 Gerson Fernando Budke
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief Atmel SAM0 MCU family I/O Pin Controller (PORT)
+ */
+
+#include "soc_gpio.h"
+
+int soc_gpio_pinmux_set(PortGroup *pg, uint32_t pin, uint32_t func)
+{
+	bool is_odd = pin & 1;
+	int idx = pin / 2U;
+
+	/* Each pinmux register holds the config for two pins.  The
+	 * even numbered pin goes in the bits 0..3 and the odd
+	 * numbered pin in bits 4..7.
+	 */
+	if (is_odd) {
+		pg->PMUX[idx].bit.PMUXO = func;
+	} else {
+		pg->PMUX[idx].bit.PMUXE = func;
+	}
+	pg->PINCFG[pin].bit.PMUXEN = 1;
+
+	return 0;
+}
+
+void soc_gpio_configure(const struct soc_gpio_pin *pin)
+{
+	PortGroup *pg = pin->regs;
+	uint32_t flags = pin->flags;
+	uint32_t func =  (pin->flags & SOC_GPIO_FUNC_MASK) >> SOC_GPIO_FUNC_POS;
+
+	/* Reset or analog I/O: all digital disabled */
+	pg->PINCFG[pin->pinum].reg = 0;
+	pg->DIRCLR.reg = (1 << pin->pinum);
+	pg->OUTCLR.reg = (1 << pin->pinum);
+
+	if (flags & SOC_GPIO_PMUXEN_ENABLE) {
+		soc_gpio_pinmux_set(pg, pin->pinum, func);
+		return;
+	}
+
+	if (flags & (SOC_GPIO_PULLUP | SOC_GPIO_PULLDOWN)) {
+		if (flags & SOC_GPIO_PULLUP) {
+			pg->OUTSET.reg = (1 << pin->pinum);
+		}
+
+		pg->PINCFG[pin->pinum].bit.PULLEN = 1;
+	}
+
+	if (flags & SOC_GPIO_INPUT_ENABLE) {
+		pg->PINCFG[pin->pinum].bit.INEN = 1;
```

Comment:
The bit access to the registers is not very efficient. It results in read, modify, write operation. However we access `PINCFG` 3 times, it's probably not worth the effort to change that.